lib/scheduler.ml

```diff
@@ -428,23 +430,44 @@ module Make (Streamd : StreamDescriptor) = struct
           (* We check for activity again, because the stream may have gone
            * inactive after the call to `write` above. *)
           written, Streamd.requires_output descriptor
-        else (
-          match PriorityQueue.pop p.children with
-          | Some ((id, (Stream i as i_node)), children') ->
-            p.t_last <- i.t;
-            let written, subtree_is_active = schedule i_node in
-            if subtree_is_active
-            then (
-              update_t i_node written;
-              p.children <- PriorityQueue.add id i_node children')
-            else (
-              implicitly_close_idle_stream i.descriptor max_seen_ids;
-              p.children <- children');
-            written, subtree_is_active
-          | None ->
-            (* Queue is empty, see line 6 above. *)
-            0, false)
+        else traverse p_node children
+    and traverse : type a. a node -> PriorityQueue.t -> int * bool =
+     fun p_node ->
+      let rec loop children =
+        match PriorityQueue.pop children with
+        | Some ((id, (Stream i as i_node)), children') ->
+          let written, subtree_is_active = schedule i_node in
+          if not subtree_is_active
+          then (
+            implicitly_close_idle_stream i.descriptor max_seen_ids;
+            (* XXX(anmonteiro): we may not want to remove from the tree right
+             * away. *)
+            remove_child p_node id);
+          if written > 0
+             (* We need this check because the queue contains both streams that
+              * want to send data and streams that don't want to send data.
+              * Without this check, we might end up in a situation where a stream
+              * won't be flushed until a few write operations later. *)
+          then (
+            update_t_last p_node i.t;
```

Comment:
we were previously updating the `t_last` of the parent node to the `t_last` of the stream that was popped from the pq. Now we only do that if we wrote something. Is that intentional?

Reply:
I feel like this is contrary to the algorithm, where it seems like it  updates `t_last` to the popped stream even if it's not going to write anything (though this doesn't seem to be very clear to me)?

Reply:
Some more reasoning that leads me to conclude that your code might be correct as is:

- if we're traversing the tree until we write something, the `t_last` for the parent would be dictated by the stream that eventually wins (writes something). 
  - in this case, it doesn't matter whether we update the `t_last` of the parent node or not
- if we end up not writing anything, we should keep `t_last` of the stream that last wrote something

Reply:
This is quite tricky and I'm not 100% convinced I'm right, but my reasoning was like yours above.
What makes it difficult to reason about is that the tree of priority queues in the present implementation contains all streams regardless of whether they have data to send or not - cf slide 14 where the assumption is that, in contrast to the dependency tree, it only contains the active streams.